Add documentation on OVN BGP agent setup with EVPN (#51)
SovereignCloudStack/hardware-landscape: documentation/sdn/ovn_bgp_evpn_deployments.md

# OVN BGP Agent in Kolla-Ansible for scalability and cloud interconnect deployments

## Test deployment of OVN BGP Agent in Kolla-Ansible
This part is a quick how-to for setting up OVN BGP Agent in Kolla-ansible.

> There's an existing [PR](https://review.opendev.org/c/openstack/kolla-ansible/+/891622) in kolla-ansible upstream, however it does not work properly. I created a [PR in kolla-ansible fork](https://github.com/SovereignCloudStack/kolla-ansible/pull/5) as well as in [kolla fork](https://github.com/SovereignCloudStack/kolla/pull/2) that attempts to fix the issues.

Reviewer: ```suggestion
There's an existing [PR](https://review.opendev.org/c/openstack/kolla-ansible/+/891622) in kolla-ansible upstream, however, it does not work properly. To fix the issues, a [PR in kolla-ansible fork](https://github.com/SovereignCloudStack/kolla-ansible/pull/5) and a [PR in kolla fork](https://github.com/SovereignCloudStack/kolla-ansible/pull/2) have been created in an attempt to fix the issues.
```

I got all containers running, however they do not seem to activate  (see below)

### Steps

> - Get a working test environment i.e. testbed, cloud-in-a-box, [kolla-builder](https://github.com/SovereignCloudStack/kolla-builder)

Reviewer: ```suggestion
- Set up a functional test environment, such as an OSISM [test environment](https://osism.tech/docs/guides/deploy-guide/examples/) like `Testbed` or `Cloud-in-a-Box`. Alternatively, you can streamline test environment creation using the [kolla-builder](https://github.com/SovereignCloudStack/kolla-builder) project, which this tutorial will demonstrate.
```

 - If you use kolla-builder, you need to checkout a dev branch `ovn-bgp-agent-deployment`
 ```bash
 git fetch origin ovn-bgp-agent-deployment:ovn-bgp-agent-deployment

>  git checkout ovn-bgp-agent-deployment

Reviewer: Instead of the above, you can use just 
```bash
 git clone --branch ovn-bgp-agent-deployment https://github.com/SovereignCloudStack/kolla-builder.git
```

```yaml
inventory_name: "ovn-bgp"
image_pool: "/home/matus/pool"
image_path: "{{ image_pool }}/kolla-image.qcow2"
ara_enable: false
is_remote: false
python_bin_path: "/home/matus/scs/kolla-venv/bin"
# Use different CIDRs and virbrX interfaces if
# the provided ones are already in use
network_ssh: "kolla-ssh-bgp-flat"
network_openstack: "kolla-openstack-bgp-flat"
network_neutron: "kolla-neutron-bgp-flat"
network_ssh_ip: "192.168.122.1"
network_openstack_ip: "192.168.123.0"
network_neutron_ip: "192.168.100.0"
network_ssh_bridge: virbr0
network_openstack_bridge: virbr1
network_neutron_bridge: virbr2
create_networks: true
kolla_internal_vip_address: "192.168.123.200"
git_patch_kolla: true
git_branch: "ovn-bgp-agent"
git_patches: []
kolla_ssh_key: "{{ lookup('ansible.builtin.env', 'HOME') }}/.ssh/id_kolla"
ssh_config: "ssh_config"
globals_file: user-globals/ovn-bgp.yml
kolla_source_local_path: "<path to kolla ansible source code>"
is_aio: true
node_name: "kolla-ovn-bgp-flat"
ram_size: 10000
cpu_count: 12

```
- Clone  [kola fork](https://github.com/SovereignCloudStack/kolla)
```
git clone git@github.com:SovereignCloudStack/kolla.git
```

> >Note: the  steps regarding building kolla images are recommended the all-in-one VM spawned by kolla-builder rather than your local machine to simplyfy deployment.

Reviewer: ```suggestion
>Note: It is recommended to follow the steps for building Kolla images on the all-in-one VM created by kolla-builder, rather than on your local machine, to simplify the deployment process.
```

> - Clone  [kola fork](https://github.com/SovereignCloudStack/kolla)

Reviewer 2: No feedback, fixed by myself.

>- run `./pre-deploy` script on the VM (this will install docker)

> >- repeat steps ommitting pushing images as it is not necessary, them being already there

Reviewer: ```suggestion
>- Repeat the steps, excluding the image-pushing step, as the images are already available
```

kolla/neutron-base                                  17.1.0           0bb7236a857f   11 days ago    1.53GB
kolla/openstack-base                                17.1.0           a498ab05494c   11 days ago    1.17GB
kolla/horizon                                       17.1.0           c694f5bda748   12 days ago    1.4GB
kolla/ovn-bgp-agent                                 17.1.0           822880a9ec91   12 days ago    1.58GB
kolla/frr                                           17.1.0           8b6a4226d03e   12 days ago    672MB
kolla/base                                          17.1.0           cbdec5ad8b29   12 days ago    631MB
```

- Push images
```
# Example commands
docker image tag kolla/neutron-server:17.1.0  registry-host:5000/myadmin/kolla/neutron-server:latest
docker image push registry-host:5000/myadmin/kolla/neutron-server:latest
# Do this for every image that was built
```
>Note: as said above, skip this step if you built your images on kolla-builder all-in-one VM

- With your images ready you can clone [kolla ansible fork](https://github.com/SovereignCloudStack/kolla-ansible)
```

> git clone git@github.com:SovereignCloudStack/kolla-ansible.git

Reviewer: ```suggestion
git clone --branch ovn-bgp-agent git@github.com:SovereignCloudStack/kolla-ansible.git
```

```yaml
additional_uploads_deploy_node:
- "<path to ovn bgp agent source>"
additional_upload_node_path: "/home/kolla"
# This is important for
additional_upload_mode: "0777"
```
- Then mount the source code by adding this to `globals.yml`
```yaml
default_extra_volumes:
- "/home/kolla/ovn-bgp-agent/ovn_bgp_agent:/var/lib/kolla/venv/lib/python3.9/site-packages/ovn_bgp_agent"
```

## Example deployments

### NB/SB Driver BGP advertising

#### Steps

- Spawn a test VM with `frr` installed `sudo apt install frr` configure bgp peering
```
frr version 9.1.2
frr defaults traditional
hostname <your VM hostname>
log file /var/log/frr/frr.log
!
router bgp 65001
 bgp router-id 1.1.1.0
 bgp log-neighbor-changes
 no bgp ebgp-requires-policy
 no bgp default ipv4-unicast
 bgp graceful-shutdown
 neighbor kolla peer-group
 neighbor kolla remote-as 65001
 neighbor kolla remote-as internal
 neighbor <INTERFACE THAT IS ON THE SAME NETWORK AS KOLLA> interface peer-group kolla
 neighbor <IP ADRESS OF KOLLA AIO VM> peer-group kolla
 !
 address-family ipv4 unicast
  neighbor kolla activate
 exit-address-family
 !
 address-family ipv6 unicast
  neighbor kolla activate
 exit-address-family
exit
```

> - Deploy kolla-ansible **without** OVN BGP agent on. This will create necessary to create ovs bridges.

Reviewer 2: Might be " This will create the necessary OVS bridges." better?

Reviewer 2: No feedback, fixed by myself.

bgp_expose_tenant_networks: "yes"
bgp_local_asn: "64999"
bgp_loopback_ip: "172.30.1.1"
# Ip of a machine you want to advertise bgp routes to
frr_neighbor_ip: "<IP ADRESS OF TEST VM>"
```
- Deploy kolla-ansible again, with new globals.
```bash
kolla-ansible -i all-in-one deploy --tags ovn-bgp-agent,frr
```
- New interfaces will be added  `bgp-vrf` and `bgp-nic`
- Adding new floating IPs, rounters or VMs to provider network will expose them by adding them to `bgp-nic`
```bash
ip a show bgp-nic
# Example output
9: bgp-nic: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue master bgp-vrf state UNKNOWN group default qlen 1000
    link/ether 16:23:48:49:bb:a6 brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.186/32 scope global bgp-nic
       valid_lft forever preferred_lft forever
    inet 192.168.100.194/32 scope global bgp-nic
       valid_lft forever preferred_lft forever
    inet 192.168.100.187/32 scope global bgp-nic
       valid_lft forever preferred_lft forever
```

### Notes/TODO

- There seems to be an error in ovn-bgp-agent, it's described in this [launchpad](https://lists.launchpad.net/yahoo-eng-team/msg94358.html)

    - This error can be avoided but it requires some hacking

>     - it seems that the externam MAC is in `external_ids` field

Reviewer: ```suggestion
    - it seems that the external MAC is in `external_ids` field
```
